packages/node/src/integrations/tracing/nest/sentry-nest-instrumentation.ts

```diff
@@ -163,33 +163,84 @@ export class SentryNestInstrumentation extends InstrumentationBase {
-                        const originalHandle = Reflect.get(thisArgNext, property, receiver);
-                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                        return (...args: any[]) => {
-                          span.end();
-
-                          if (prevSpan) {
-                            return withActiveSpan(prevSpan, () => {
-                              return Reflect.apply(originalHandle, thisArgNext, args);
-                            });
-                          } else {
-                            return Reflect.apply(originalHandle, thisArgNext, args);
+                const parentSpan = getActiveSpan();
+                let afterSpan: Span;
+
+                return startSpanManual(getMiddlewareSpanOptions(target), (beforeSpan: Span) => {
+                  // eslint-disable-next-line @typescript-eslint/unbound-method
+                  next.handle = new Proxy(next.handle, {
+                    apply: (originalHandle, thisArgHandle, argsHandle) => {
+                      beforeSpan.end();
+
+                      if (parentSpan) {
+                        return withActiveSpan(parentSpan, () => {
+                          const handleReturnObservable = Reflect.apply(originalHandle, thisArgHandle, argsHandle);
+
+                          if (!request._sentryInterceptorInstrumented) {
+                            afterSpan = startInactiveSpan(
+                              getMiddlewareSpanOptions(target, 'Interceptor - After Route'),
+                            );
                           }
-                        };
-                      }
 
-                      return Reflect.get(target, property, receiver);
+                          return handleReturnObservable;
+                        });
+                      } else {
+                        const handleReturnObservable = Reflect.apply(originalHandle, thisArgHandle, argsHandle);
+
+                        if (!request._sentryInterceptorInstrumented) {
+                          afterSpan = startInactiveSpan(getMiddlewareSpanOptions(target, 'Interceptor - After Route'));
+                        }
+
+                        return handleReturnObservable;
+                      }
                     },
                   });
 
-                  return originalIntercept.apply(thisArgIntercept, [executionContext, nextProxy, args]);
+                  let returnedObservableInterceptMaybePromise: Observable<unknown> | Promise<Observable<unknown>>;
+
+                  try {
+                    returnedObservableInterceptMaybePromise = originalIntercept.apply(thisArgIntercept, argsIntercept);
+                  } catch (e) {
+                    if (!request._sentryInterceptorInstrumented) {
+                      if (beforeSpan) beforeSpan.end();
+                      if (afterSpan) afterSpan.end();
+                      addNonEnumerableProperty(request, '_sentryInterceptorInstrumented', true);
```

Comment:
h: I think we have a race condition with the non-enumerable property. We to also set the flag before we start the span, otherwise we might start multiple if multiple async observers run in parallel.

Reply:
Jep. Setting the property immediately now and updated the rest of the implementation to not rely on the property, should be fixed